Content.Server/Chemistry/Containers/EntitySystems/SolutionContainerSystem.cs

```diff
@@ -0,0 +1,196 @@
+            };
+            AddComp(solutionId, solutionComp);
+            existed = false;
+        }
+        else
+        {
+            solutionComp.Solution.MaxVolume = FixedPoint2.Max(solutionComp.Solution.MaxVolume, minVol);
+            Dirty(solutionId, solutionComp);
+        }
+
+        if (!TryComp(solutionId, out ContainerSolutionComponent? relation))
+        {
+            relation = new ContainerSolutionComponent() { Container = uid, Name = name };
+            AddComp(solutionId, relation);
+        }
+        else
+        {
+            (relation.Container, relation.Name) = (uid, name);
+            Dirty(solutionId, relation);
+        }
+
+        if (needsInit)
+            EntityManager.InitializeAndStartEntity(solutionId, Transform(solutionId).MapID);
+
+        return (solutionId, solutionComp);
+    }
+
+    private Solution EnsureSolutionPrototype(Entity<SolutionContainerManagerComponent?> entity, string name, FixedPoint2 minVol, out bool existed)
+    {
+        existed = true;
+
+        var (uid, container) = entity;
+        if (!Resolve(uid, ref container, logMissing: false))
+        {
+            container = AddComp<SolutionContainerManagerComponent>(uid);
+            existed = false;
+        }
+
+        if (!container.Solutions.TryGetValue(name, out var prototype))
+        {
+            prototype = new Solution() { MaxVolume = minVol, Name = name };
+            container.Solutions.Add(name, prototype);
+            existed = false;
+        }
+        else
+            prototype.MaxVolume = FixedPoint2.Max(prototype.MaxVolume, minVol);
+
+        return prototype;
+    }
+
+
+    private Entity<SolutionComponent> SpawnSolutionUninitialized(EntityUid container, string name, Solution prototype)
+    {
+        var coords = new EntityCoordinates(container, Vector2.Zero);
+        var uid = EntityManager.CreateEntityUninitialized(null, coords, null);
+        if (!TryComp(uid, out SolutionComponent? comp))
+        {
+            var solution = prototype.Clone();
+            solution.Name = name;
+            comp = new SolutionComponent() { Solution = solution };
+            AddComp(uid, comp);
+        }
+
+        if (!TryComp(uid, out ContainerSolutionComponent? relation))
+        {
+            relation = new ContainerSolutionComponent() { Container = container, Name = name };
+            AddComp(uid, relation);
+        }
+        else
+            (relation.Container, relation.Name) = (container, name);
+
+        return (uid, comp);
+    }
+
+
+    #region Event Handlers
+
+    private void OnMapInit(EntityUid uid, SolutionContainerManagerComponent comp, MapInitEvent args)
+    {
+        if (comp.Solutions is { Count: > 0 } prototypes)
+        {
+            foreach (var (name, prototype) in prototypes)
+            {
+                var solution = EnsureSolutionEntity(uid, name, prototype.MaxVolume, out _);
+                SolutionSystem.AddSolution(solution, prototype);
+            }
```

Comment:
I'd say just re-use the existing solution object instead of creating a new one and adding its contents, because solution addition isn't all that fast and this migh slow down map initialization.